Extract ErrorMatcherMixin from Catch and Retry

kbrock opened the pull request:
- Extract ErrorMatcherMixin#match_error?
- `States.HeartbeatTimeout` is handled by `States.Timeout`
- `States.Runtime` is not caught by any catchers/retriers

Fryguy commented:
I think your branch name and PR title are wrong?

Fryguy commented on lib/floe/workflow/error_matcher.rb:
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module Floe
+  class Workflow
+    module ErrorMatcher
+      # note, expecting error_equals from payload to be available
```
Since this comment is a requirement, I wouldn't call it a note - I think this can more clearly be written as a comment on the module:


```suggestion
module Floe
  class Workflow
    # Contains methods for common error handling
    #
    # Requires that the including class to have the `error_equals` from the payload to be available.
    module ErrorMatcher
      # note, expecting error_equals from payload to be available
```

or something like that.

Fryguy replied:
Or perhaps it's just an addition to the comment for the `match_error?` method specifically.

Fryguy replied:
Or perhaps it's a note, but using `@note` yard tag in the `match_error?` method's comment specifically.  I can't really find an `@requires` or `@depends` yard tag.

kbrock replied:
I was able to extract this into the parent so this is no longer needed.

kbrock replied:
that was too much of a leap.
I went back to keeping this PR as just extraction and moving that into #212

kbrock commented:
update:
- fixed note
- fixed specs (was not properly comparing `equals_value`)

kbrock commented:
un-wip: tests passing

Fryguy commented on lib/floe/workflow/error_matcher.rb:
```diff
@@ -0,0 +1,19 @@
+      # @param [String] error the error thrown
+      def match_error?(error)
+        return false if error == "States.Runtime"
+        return true if error_equals.include?("States.ALL")
```
Do we need to check that the author set error_equals correctly?  If it was unset, then I think this is nil, and if not set to an Array, this is a string include check?  (I think the old code also failed, but it was more subtle).

Should we add validation of the error equals payload?

kbrock replied:
old code was:

```ruby
(r.error_equals & [error, "States.ALL"]).any?

nil & [1,2]
# => nil
```

So yea, it silently failed.

kbrock replied:
Added validation so we catch a retry with an empty, missing, or non-array ErrorEquals

kbrock commented:
update:
- added error_equals validation

kbrock commented:
update:
- moved error_equals up to parent - no longer the `Note` stuff
- added "States.ALL" special case
- generalized Heartbeat special case
- specs for `ErrorMatcher`

kbrock commented on lib/floe/workflow/error_matcher.rb:
```diff
@@ -0,0 +1,34 @@
+        validate_error_equals!(name, ".#{payload["Type"]}")
+
+        @error_equals_all = error_equals.include?("States.ALL")
+        error_equals << "States.HeartbeatTimeout" if error_equals.include?("States.Timeout")
```
I kept this optimization in a separate commit.

2 camps:
- modify this array and no special `if` cases.
- don't modify this array and add special `if` logic.

kbrock commented:
update:

- rewrote tests to go through workflow
- added tests for validation
- now checks that retrier (and catch) are non-empty

update:

- fixed `tap` in tests

kbrock commented:
update:
- fix code complexity via validate_array_field!

kbrock commented:
update:

- compressed _catch_all case to convert guard into just a `boolean ||`

kbrock commented:
update:
- rebased
- depend upon common error handling

kbrock commented:
update:
- rebase (fixed conflict)
- rubocop: fixed spacing

kbrock commented:
update:
- based upon #209 
- Greatly simplified the logic in here.

kbrock commented:
update:
- remove 209 - trimmed it down

kbrock commented:
update:
- rubocop

kbrock commented:
update:
- extracted #212 (few pushes back) to make this as simple as I can

update:
- rebase master

kbrock commented:
@agrare @Fryguy this has been greatly simplified. Is this good to go now?
